### ISSUE TITLE
SHACL rule-based reasoner vs Jena reasoner: inferring on new triples

### ISSUE THREAD

**Issue by SonyaSt** [TopQuadrant/shacl#65]
I'm trying to implement SHACL inference engine and I have remarked that one scenario is not working as expected. 
And it is supported by Jena rule engine. 
In fact, Jena manages cases when, after first inference, rules need to be applied once again on inferred triples. It is a sort of looping till there is no more data to infer.
I have an impression that this is not the case for SHACL engine.

Here is my Ontology: 
```
ex:Viewer 
	rdf:type ex:Role ;
    ex:accessTo ex:Read ;
.

ex:Editor 
	rdf:type ex:Role ;
	ex:accessTo ex:Edit ;
	ex:includes ex:Viewer ;
.

ex:Admin 
	rdf:type ex:Role ;
    ex:accessTo ex:Administrate ;
	ex:includes ex:Editor ;
.

ex:Role 
	rdf:type rdfs:Class;
.
```

Here the SHACL rule:
```
ex:permissionShape
	rdf:type sh:NodeShape ;
	sh:targetClass ex:Role ;
	sh:rule [
	  rdf:type sh:SPARQLRule ;
	  sh:construct """
			CONSTRUCT {
				$this ex:accessTo ?action .
			}
			WHERE {
				$this ex:includes ?role .
				?role ex:accessTo ?action .
				FILTER NOT EXISTS { $this ex:accessTo ?action . }
			}
			""" ;
	  sh:prefixes ex: ;
	] ;
.
```

I'm expecting that following triples get generated for ex:Admin:
ex:Admin ex:accessTo ex:Edit (through  ex:includes of ex:Editor)
ex:Admin ex:accessTo ex:Read (through  ex:Editor that ex:includes ex:Viewer)

But only the first one gets generated for ex:Admin.

This works as expected with Jena rule engine, both triples are generated.
Jena rule:
`[ rule1: ( ?a ex:includes ?b ), ( ?b ex:accessTo ?action )
 -> ( ?a ex:accessTo ?action ) ]`
 
Is there a possibility to activate this kind of "recursion" in SHACL inference engine?

**Comment by simonstey** [comment]
> In fact, Jena manages cases when, after first inference, rules need to be applied once again on inferred triples. It is a sort of looping till there is no more data to infer.
**I have an impression that this is not the case for SHACL engine.**

which is complaint with [8.4 General Execution Instructions for SHACL Rules](https://www.w3.org/TR/shacl-af/#rules-execution):

>In order to count as a SHACL rules engine, an implementation must be capable of inferring triples according to the following procedure (given in pseudo-code), or a different algorithm as long as the result is the same as specified. **Note that this algorithm only covers a single "iteration" over all rules, without prescribing the behavior if the same rule needs to be applied multiple times after other rules have fired.** The latter is left to future work.
```
for each shape S in the shapes graph, ordered by execution order {
		for each non-deactivated rule R in the shape, ordered by execution order {
			for each target node T of S that conforms to all conditions of R {
				execute R using T as focus node following the execution instructions of its rule types
			}
		}
	}
```
> **The triples that are inferred by a rule do not immediately become part of the data graph, i.e. the triples produced by one rule can not always be queried by other rules.** These policies reduce the likelihood of race conditions and better support parallel execution.
>
> - If two shapes have the same execution order then their newly inferred triples are not visible to each other.
> - If two rules have the same execution order then their newly inferred triples are not visible to each other.
> - If the same rule is executed on multiple target nodes then the newly inferred triples are not visible to the other target nodes.

you could implement something like that yourself, though ;)

**Comment by SonyaSt** [comment]
Thanks for a quick answer!

I have added "sh:entailment sh:Rules" to my Ontology.
And now both triples are generated! 
ex:Admin ex:accessTo ex:Edit 
ex:Admin ex:accessTo ex:Read

Curiouser and curiouser! :)

Do you know why? Is this a good way to proceed?

**Comment by HolgerKnublauch** [comment]
As Simon states, the default implementation is to just do a single iteration - each rule/focus node is executed once. The trick with sh:entailment sh:Rules covers two iterations: one to prepare the input graph and another to do the usual step. So I don't think it would help inferences that are more than two steps away from another inferred triple.

To implement this in a more generic way, just call RuleEngine.executeAll() until the size of the inferences graph has not increased. The idea is that rules are then repeatedly executed until a fixpoint is reached and no further inferences have happened during one round.

**Comment by SonyaSt** [comment]
Thanks, that worked!

**Comment by mrtnzagustin** [comment]
I think i have a design problem, but is possible to run a subset of rules to make some Class inference, and after that execution, run another rule that use that firsts inferences? My "second line" rules need the firsts rules to be executed before.

Example:
1) First Level Rule: 
```sparql
CONSTRUCT { ?individual rdf:type myOntoIri:InfClass } 
WHERE { ... }
```
2) Second Level Rule: 
```sparql
CONSTRUCT {?indInfClass ....} 
WHERE { ... 
   FILTER NOT EXISTS ( 
      SELECT ?infIndClass 
      WHERE { ?infIndClass rdf:type myOntoIri:InfClass} 
}
```
In my tests, the second level rules is being executed in parallel with the first ones and all my individuals get type myOntoIri:InfClass assigned. I am using sh:order 0 and 1 for each subset of rules.
What i am doing wrong?

PD: Sorry for my english.

Regards!!

**Comment by HolgerKnublauch** [comment]
It should work as you describe, assuming that the query model passed into the Rule engine includes the inferences graph. It would first execute the order 0 rules and then order 1 rules, which would normally see the inferences from the previous step. I guess you'd need to provide more details about how you are calling this and the exact shape definitions.

**Comment by mrtnzagustin** [comment]
@HolgerKnublauch, an update to this.
I was doing this:
```sparql
:firstLevelRule a sh:NodeShape ;
    sh:targetClass :Ingreso;
	sh:rule [ ... ] . 
:secondLevelRule a sh:NodeShape ;
    sh:targetClass :Ingreso;
	sh:rule [ ... ] . 
:thirdLevelRule a sh:NodeShape ;
    sh:targetClass :Ingreso;
	sh:rule [ ... ] . 
```
And now i understand that all rules have to be in the same shape!
```sparql
:AllRules a sh:NodeShape ;
    sh:targetClass :MyClass;
	sh:rule :firstLevelRule ;
	sh:rule :secondLevelRule ;
	sh:rule :thirdLevelRule .

:firstLevelRule a sh:SPARQLRule ;
 .... ;
:secondLevelRule a sh:SPARQLRule ;
 .... ;
:thirdLevelRule a sh:SPARQLRule ;
 .... ;
```
Now its working, thank you very much!!
